include/seqan3/container/detail/random_access_iterator.hpp

```diff
@@ -0,0 +1,235 @@
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file container/detail/random_access_iterator.hpp
+ * \brief Random access iterator for const and non const containers.
+ * \author Marie Hoffmann <marie.hoffmann AT fu-berlin.de>
+ * \ingroup iterator
+ */
+
+#pragma once
+
+#include <iomanip>
+#include <limits>
+#include <type_traits>
+#include <cassert>
+
+#include <range/v3/all.hpp>
+
+#include <seqan3/core/concept/iterator.hpp>
+#include <seqan3/range/concept.hpp>
+
+namespace seqan3::detail
+{
+
+template <typename container_type>
+    requires random_access_range_concept<container_type> && sized_range_concept<container_type>
+class random_access_iterator
+{
+
+private:
+    //! \brief Iterator stores pointer to underlying container structure
+    typename std::add_pointer_t<container_type> host{nullptr};
+    using position_type =  ranges::v3::size_type_t<container_type>;
+    //! \brief Store position index for container
+    position_type pos{static_cast<position_type>(0)};
+
+public:
+    //! \brief Type for distances between iterators
+    using difference_type = ranges::v3::difference_type_t<container_type>;
+    //! \brief Value type of container elements
+    using value_type = ranges::v3::value_type_t<container_type>;
+    //! \brief Reference types defined by container
+    using reference = typename container_type::reference;
+    using const_reference = typename container_type::const_reference;
```

Comment:
this needs a documentation entry, as well! Please run doxygen and grep for your files so that you find any undocumented entries!